CP-47001 Introduce file descriptor testing framework

edwintorok opened the pull request:
This will be used to test the stdext and XAPI epoll changes.

I'm breaking this up into one module / commit to make it easier to review.

This PR is for the build plumbing and detecting double close (which will be needed to write a reliable test suite).

Initially I also had some code here that would close leaked file descriptors, but I dropped it because even with all the safety checks I'm not sure it can be done safely (you could have leaked a `Unix.file_descr` and have another thread writing to it, and if you close it then the FD could be reused by something else).

Next will be some wrappers that track read and write permissions using phantom types, and then a QuickCheck generator and property tester.

edwintorok committed:
CP-47001: [xapi-fdcaps]: dune plumbing for new library

This will be a new library that will provide a more type-safe interface to file descriptor operations.
Useful on its own, but also for testing stdext.

Minimal dependencies, only Unix (and Alcotest for testing).

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fd-test]: dune plumbing for a new test framework

This will be a test framework providing QCheck generators and properties for
testing file descriptor operations.
It will try to generate:
* different kinds of file descriptors
* actual data written/read on the other end of pipes and socket pairs
* different speeds and delays on the other end to find buffering bugs
* file descriptors that are >1024 to find bugs with select

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
[maintenance]: disable implicit transitive deps

Helps catch missing dependencies in dune files.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
Fixed the CI failure, had to add the new packages to the github action workflow.

edwintorok committed:
CP-47001: [xapi-fdcaps]: add -principal flag

We are going to use type-level constraints a lot.
Try to future proof it by using the recommended compiler flag.
`ocamlc` says this about `-principal`:
> When using labelled arguments and/or polymorphic methods, this flag is required to
> ensure future versions of the compiler will be able to infer types correctly, even if internal algorithms change

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fdcaps]: optional coverage support

This is not enabled by default (but bisect-ppx is nevertheless a build-time dependency)
Usage: `make coverage`

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fdcaps]: add properties module and tests

Lightweight wrapper using polymorphic variants to track read, write, and file kind properties on file descriptors.
We only track the property at the time the file descriptor was opened.

This prevents bugs like accidentally swapping the read and write ends of a pipe,
or attempting to run an operation on a file descriptor that would alway s fail (e.g. setting a socket timeout on a pipe)

Write tests using cram-style expect tests that the operations we expect to be forbidden by this type system
are actually forbidden.
The error messages may be compiler version dependent, so only run them on OCaml 4.14.1 for now.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fdcaps]: add operations module and tests

Use the capabilities module to wrap most Unix operations needed in testing Unixext

Add a testsuite that checks that whenever the type says "never" the underlying file descriptor operation
would indeed raise an exception. This ensures that the type constraints we declare are actually correct.
The checks use unsafe operations that bypass the type layer.

Similarly check that operations that are accepted by the type system and marked as "always" in the type succeed.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
I think the properties and operations module are now ready to review, next up would be some modules that do some actual quickchecking, after a bit more plumbing.

edwintorok commented:
(removed the automated coveralls submission,that requires admin permission to set up, we can do that separately -- and we can use either coveralls or codecov)

lindig reviewed (approved):
This is extensively documented and comes with test but it is also quite intricate and difficult to review.

edwintorok committed:
CP-47001: [xapi-fdcaps] add sections to Operations documentation

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

psafont commented on dune-project:
```diff
@@ -115,6 +115,7 @@
     (alcotest :with-test)
     base-unix
     fmt
+    (bisect_ppx :build)
```
should it also have a :with-test as well?

edwintorok replied:
I'll have to try whether you can build without (it is off by default, but present in the dune as instrument declaration). It'd certainly be desirable to limit where it is needed

edwintorok replied:
Uninstalled `bisect_ppx` and building+runtest still worked (`make coverage` obviously didn't), so I change this to ':with-test'.
This is different to how bisect-ppx worked in the past where it was always required to be present at build time even when unused, so this is a nice upstream improvement.

psafont commented on lib/xapi-fdcaps/test/properties.t:
```diff
@@ -0,0 +1,23 @@
+  > EOF
+  $ ocamlfind ocamlc -package xapi-fdcaps -c t.ml
+  File "t.ml", line 2, characters 40-42:
+  2 | let _ = as_readable (make `wronly `reg) in ()
```
Would it be an option to grep for Syntax error so it can be compiler independent?

edwintorok replied:
Actually that is the wrong error message, thanks for spotting.
The real error is something longer like:
```
Error: This expression has type
           ([< Xapi_fdcaps.Properties.rw > `rdonly ] as 'a,
            [< Xapi_fdcaps.Properties.kind > `reg ] as 'b)
           Xapi_fdcaps.Properties.t = ('a, 'b) Xapi_fdcaps.Properties.props
         but an expression was expected of type
           ([< Xapi_fdcaps.Properties.writable ] as 'c,
            [< Xapi_fdcaps.Properties.kind ] as 'd)
           Xapi_fdcaps.Properties.t = ('c, 'd) Xapi_fdcaps.Properties.props
         Type
           [< Xapi_fdcaps.Properties.rw > `rdonly ] =
             [< `rdonly | `rdwr | `wronly > `rdonly ]
         is not compatible with type
           [< Xapi_fdcaps.Properties.writable ] = [< `rdwr | `wronly ]
         The second variant type does not allow tag(s) `rdonly
```

I pushed a fixup

edwintorok replied:
Maybe I could pipe the output to tail and match only on the last line? That is unlikely to change between versions

psafont replied:
that might work, yeah

edwintorok replied:
I used `tail -n 1` now, and removed the OCaml version constraint from the tests.

edwintorok replied:
I also tried to use `set -o pipefail` but that failed on the CI, probably due to a non-bash shell. Anyway we've got the last line of the error output, so if it suddenly turns out that there is no error then that will change and cause a test failure, so it should be fine without pipefail in this case

psafont commented on lib/xapi-fdcaps/operations.ml:
```diff
@@ -0,0 +1,206 @@
+  let finally () = close t in
+  Fun.protect ~finally (fun () -> f t)
+
+module Syntax = struct let ( let@ ) f x = f x end
+
+open Syntax
```
```suggestion
let ( let@ ) f x = f x
```

edwintorok replied:
Syntax is in a separate module in the mli, you mean I should move the let declaration to the toplevel?
I thought to avoid clashes in the toplevel module (although it is unlikely that we'd define `let@` with a different semantics, clashes are more likely for the monadic let versions)

psafont commented on lib/xapi-fdcaps/operations.mli:
```diff
@@ -0,0 +1,210 @@
+val with_fd2 : 'a t * 'b t -> ('a t * 'b t -> 'c) -> 'c
+(** [with_fd2 fd1 fd2 f] calls [f fd1 fd2] and always closes [t] after [f] finishes. *)
+
+module Syntax : sig
+  val ( let@ ) : ('a -> 'b) -> 'a -> 'b
+  (** [let@ fd = with_fd t in ... use fd] *)
+end
```
I don't this module should be exposed, seems like an implementation detail. If users want to use let@ they can easily define it

edwintorok replied:
It will be used by other modules that are not part of this PR yet.

edwintorok replied:
I've kept this as is for now. `Operations` isn't the ideal place for this `let@` definition, but we should find some shared place for it to encourage its use (it is the preferred way of using the `with_fd` and `with_fd2` functions in this module

edwintorok replied:
Should I move it to a toplevel `syntax.mli` in `xapi-fdcaps`?

psafont commented on lib/xapi-fdcaps/operations.ml:
```diff
@@ -0,0 +1,206 @@
+  | _ ->
+      failwith (Printf.sprintf "%s exited nonzero" cmd)
+
+let with_temp_blk ?(sector_size = 512) ?delay_read_ms:_ ?delay_write_ms:_ name f
```
These delays should be respected, or removed from the interface

edwintorok replied:
The delays can be implemented with [dmsetup delay ](https://docs.kernel.org/admin-guide/device-mapper/delay.html). I can remove `with_temp_blk` from here and add it back once that is implemented

edwintorok committed:
CP-47001: [xapi-fdcaps]: wrap more Unix operations

edwintorok committed:
fixup! CP-47001: [xapi-fdcaps]: add properties module and tests

edwintorok committed:
fixup! CP-47001: [xapi-fdcaps]: optional coverage support

edwintorok committed:
fixup! CP-47001: [xapi-fdcaps]: add operations module and tests

edwintorok commented:
I think these modules are now mostly complete, I'll squash the fixups into the appropriate commits once you had a chance to take another look at them.
There are a few more modules to add before this can be used for quickcheck, but that will go into the xapi-fd-test package instead and I'll do that in a separate PR.

edwintorok committed:
CP-47001: [xapi-fdcaps] runtime tests for read-write properties

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fdcaps-test]: add observations module

It can be used to wrap read or write operations andobserve the data that is transferred,
and elapsed time.

It also provides 2 functions that create a file of a given kind.

We only test UNIX sockets, because socketpair doesn't support TCP sockets on Linux.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
CP-47001: [xapi-fdcaps-test]: add generate module

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
This is continued in https://github.com/xapi-project/xen-api/pull/5402
